rust/kernel/fwnode.rs

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Unified device property interface.
+//!
+//! C header: [`include/linux/property.h`](../../../../include/linux/property.h)
+
+use crate::{bindings, device, ARef, AlwaysRefCounted};
+use core::{cell::UnsafeCell, ptr};
+
+/// Represents a `struct fwnode_handle *` part of a device's fwnode graph
+///
+/// # Invariants
+///
+/// The pointer is valid.
+#[repr(transparent)]
+pub struct Node(pub(crate) UnsafeCell<bindings::fwnode_handle>);
+
+// SAFETY: The type invariants guarantee that `Node` is always ref-counted.
+unsafe impl AlwaysRefCounted for Node {
+    fn inc_ref(&self) {
+        // SAFETY: The existence of a shared reference means that the refcount is nonzero.
+        unsafe { bindings::fwnode_handle_get(self.0.get()) };
+    }
+
+    unsafe fn dec_ref(obj: core::ptr::NonNull<Self>) {
+        // SAFETY: The safety requirements guarantee that the refcount is nonzero.
+        unsafe { bindings::fwnode_handle_put(obj.cast().as_ptr()) }
+    }
+}
+
+impl Node {
+    /// Creates a new DeviceNode instance from an existing [`device::Device`] instance.
+    pub fn from(dev: &device::Device) -> Option<ARef<Self>> {
+        // SAFETY: By the type invariants, `dev` owns a reference, so it safe to access `ptr`.
+        // `fwnode_handle_get` increments the refcount of the `fwnode_handle` if it is not
+        // NULL or returns NULL.
+        let node = unsafe { bindings::fwnode_handle_get(bindings::dev_fwnode(dev.ptr)) };
+        let ptr = ptr::NonNull::new(node)?;
+        // SAFETY: `fwnode_handle_get` increments the refcount
+        Some(unsafe { ARef::from_raw(ptr.cast()) })
+    }
+    /// Creates an instance of `NodeIterator`
+    ///
+    /// This provides an `Iterator` wrapping the internal C functionality of invoking
+    /// `fwnode_graph_get_next_endpoint`
+    pub fn endpoints(node: ARef<Self>) -> EndpointIterator {
+        EndpointIterator {
+            handle: node,
+            curr_node: None,
+        }
+    }
+}
+
+/// Implements the Iterator trait to iterate the device's endpoints given the `DeviceNode`
```

Comment:
Here and elsewhere there are still references to DeviceNode.

Reply:
My bad! I have fixed them now.